lib/services/watcher.go

```diff
@@ -153,6 +174,13 @@ func (p *resourceWatcher) runWatchLoop() {
 		}
 		if err != nil {
 			p.Log.Warningf("Restart watch on error: %v.", err)
+			if p.failureStartedAt.IsZero() {
+				p.failureStartedAt = p.Clock.Now()
```

Comment:
I'm afraid your approach could cause the strict locking mode to apply too aggressively. Consider the following sequence:
1. `runWatchLoop` starts.
2. `watch` repeatedly fails on the very first `p.Client.NewWatcher` call for a couple of minutes, longer than the `MaxStaleness` duration.
2. `failureStartedAt` is set to the first failure time point and `notifyStale` is correctly called for each failure since `MaxStaleness` was exceeded.
3. `watch` gets re-executed once again but now it manages to reach the second `select` loop (i.e. the local lock view gets refreshed with `getResourcesAndUpdateCurrent`). In my approach this also resets `failureStartedAt`.
4. Let the second `select` loop fail before `refetchC` fires, making `watch` return an error.
5. In your approach, since `failureStartedAt` still refers to the original failure, `MaxStaleness` remains exceeded and so `notifyStale` is called again even though the data are in fact not stale at this point. With `LockingModeStrict` and an intermittent connection to the auth server, this could impose an almost permanent preventive lock-out.

Reply:
Given that the scenario you described requires consecutive failures in a large time window (5min by default), I'd be OK accepting it (given that, I assume, it is highly unlikely). I think the cognitive benefits of dealing with `failureStartedAt` in a single method outweighs the cost of the corner case. Up to you.

If you decide against it, consider leaving a comment like this at runWatchLoop:

```go
// runWatchLoop runs a watch loop.
func (p *resourceWatcher) runWatchLoop() {
	for {
		/* code omitted /*
		if err != nil {
			p.Log.Warningf("Restart watch on error: %v.", err)
			if p.failureStartedAt.IsZero() {
				p.failureStartedAt = p.Clock.Now()
			}
		}
		// `p.failureStartedAt` is reset on success by resourceWatcher.watch,
		// instead of at this method, in order to minimize the length of failure
		// windows in the case of intermittent connections.
		/* code omitted /*
	}
}
```

Reply:
The scenario I describe requires a connection failure to happen once per every `RefetchPeriod` (10 minutes) which I don't think is that unlikely, especially considering the whole `LockWatcher` approach has been devised to cover for such situations, see the discussion at #7286.

Regarding any "cognitive benefits" I think it's more agreeable to keep the handling of `p.retry` and `p.failureStartedAt` together at the same points (which is what I also thought to be intuitive when writing the code).

Reply:
So the refetch period (10m) is larger than the staleness window (5m)? That sounds problematic, or am I reading it wrong?

As for the `p.failureStartedAt` change, I'm happy with a comment on runWatchLoop, as stated above.

Reply:
> So the refetch period (10m) is larger than the staleness window (5m)?

Yes.

Reply:
Won't that always trigger a failure state in the absence of other events? Shouldn't we use a default for the staleness windows that is strictly larger than the refetch period?

Reply:
I don't think so. Excluding other errors, `watch` returns no error when `refetchC` fires, ergo `failureStartedAt` remains unset, ergo the staleness condition cannot obtain.

Reply:
Got it, thanks for the explanation.

Keeping this open so we can get that comment in (or you can push back).